server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -130,47 +135,73 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
             indexResponses.putIfAbsent(resp.getIndexName(), resp);
         };
         final FailureCollector indexFailures = new FailureCollector();
-        // One for each cluster including the local cluster
-        final CountDown completionCounter = new CountDown(1 + remoteClusterIndices.size());
-        final Runnable countDown = createResponseMerger(request, fieldCapTask, completionCounter, indexResponses, indexFailures, listener);
-        final RequestDispatcher requestDispatcher = new RequestDispatcher(
-            clusterService,
-            transportService,
-            task,
-            request,
-            localIndices,
-            nowInMillis,
-            concreteIndices,
-            threadPool.executor(ThreadPool.Names.SEARCH_COORDINATION),
-            handleIndexResponse,
-            indexFailures::collect,
-            countDown
-        );
-        requestDispatcher.execute();
+        try (RefCountingRunnable refs = new RefCountingRunnable(() -> {
+            indexMappingHashToResponses.clear();
+            if (fieldCapTask.notifyIfCancelled(listener)) {
```

Comment:
Hmm AIUI this might notify the caller immediately on cancellation, even though there are some outstanding remote actions. I think we should wait for pending remote actions to complete before responding, although of course it is important to react (internally) to the cancellation right away like this.

In other words, we should acquire a ref when _sending_ each remote request, rather than trying to acquire one when we receive the response.

Reply:
Thanks David. I've updated the PR to clear the resources on cancelled, stop accumulating errors/responses after cancelled, and wait for remote actions before replying to the caller.

Reply:
:+1: looks good